auth/user_mgt.go

```diff
@@ -652,6 +593,156 @@ type userManagementClient struct {
 	httpClient *internal.HTTPClient
 }
 
+// GetUser gets the user data corresponding to the specified user ID.
+func (c *userManagementClient) GetUser(ctx context.Context, uid string) (*UserRecord, error) {
+	return c.getUser(ctx, uidQuery(uid))
+}
+
+// GetUserByEmail gets the user data corresponding to the specified email.
+func (c *userManagementClient) GetUserByEmail(ctx context.Context, email string) (*UserRecord, error) {
+	if err := validateEmail(email); err != nil {
+		return nil, err
+	}
+	return c.getUser(ctx, emailQuery(email))
+}
+
+// GetUserByPhoneNumber gets the user data corresponding to the specified user phone number.
+func (c *userManagementClient) GetUserByPhoneNumber(ctx context.Context, phone string) (*UserRecord, error) {
+	if err := validatePhone(phone); err != nil {
+		return nil, err
+	}
+	return c.getUser(ctx, phoneNumberQuery(phone))
+}
+
+type userQuery interface {
+	name() string
+	build() map[string]interface{}
+}
+
+type uidQuery string
```

Comment:
This isn't bad, but every one of these implementations is the same, except for a hard-coded string. What about this?

```
type userQuery {
	field string
	value string
}

func (q userQuery) description() string {
	return fmt.Sprintf("%s: %q", q.field, q.value)
}

func (q userQuery) build() map[string]interface{} {
	return map[string]interface{}{
		q.field: []string{string(q.value)},
	}
}
```

Then instead of like `phoneNumberQuery(phone)`, it's `userQuery{"phoneNumber", phone}`. The potential error message when it fails isn't quite as pretty, but it's a lot less code. That is, assuming there won't be other types of user queries.

I'm okay either way, but thought I'd suggest it.

Reply:
I like the simplicity of this approach, so I'm implementing it with a minor modification to retain the nicer error messages. We are unlikely to introduce more complex queries in the future. But if we do, we can implement the interface-based solution at that time.